extensions/common/iam/identity-trust/identity-trust-service/src/main/java/org/eclipse/edc/iam/identitytrust/verification/SelfIssuedIdTokenVerifier.java

```diff
@@ -0,0 +1,105 @@
+ * <p>
+ * Please note that <strong>no structural</strong> validation is done beyond the very basics (must have iss and aud claim).
+ * This is done by the {@link org.eclipse.edc.iam.identitytrust.validation.SelfIssuedIdTokenValidator}.
+ *
+ * @see org.eclipse.edc.iam.identitytrust.validation.SelfIssuedIdTokenValidator For SI Token validation.
+ */
+public class SelfIssuedIdTokenVerifier implements JwtVerifier {
+    private final DidResolverRegistry resolverRegistry;
+
+    public SelfIssuedIdTokenVerifier(DidResolverRegistry resolverRegistry) {
+        this.resolverRegistry = resolverRegistry;
+    }
+
+    @Override
+    public Result<Void> verify(TokenRepresentation tokenRepresentation, String audience) {
+
+        SignedJWT jwt;
+        try {
+            jwt = SignedJWT.parse(tokenRepresentation.getToken());
+            var didResult = resolverRegistry.resolve(jwt.getJWTClaimsSet().getIssuer());
+            if (didResult.failed()) {
+                return Result.failure("Unable to resolve DID: %s".formatted(didResult.getFailureDetail()));
+            }
+
+            // this will return the _first_ public key entry
+            var keyId = jwt.getHeader().getKeyID();
+
+            //either get the first verification method, or the one specified by the key id
+            var publicKey = Optional.ofNullable(keyId)
+                    .map(kid -> getVerificationMethod(didResult.getContent(), kid))
+                    .orElseGet(() -> firstVerificationMethod(didResult.getContent()));
+
+            if (publicKey.isEmpty()) {
+                return Result.failure("Public Key not found in DID Document.");
+            }
+
+            //convert the POJO into a usable PK-wrapper:
+            var publicKeyJwk = publicKey.get().getPublicKeyJwk();
+            var publicKeyWrapperResult = KeyConverter.toPublicKeyWrapper(publicKeyJwk, publicKey.get().getId());
+            if (publicKeyWrapperResult.failed()) {
+                return publicKeyWrapperResult.mapTo();
+            }
+
+            var verified = JwtUtils.verify(jwt, publicKeyWrapperResult.getContent(), audience);
+            if (verified.failed()) {
+                return Result.failure("Token could not be verified: %s".formatted(verified.getFailureDetail()));
+            }
+            return Result.success();
+        } catch (ParseException e) {
+            return Result.failure("Error parsing JWT");
+        }
+    }
+
+    private Optional<VerificationMethod> getVerificationMethod(DidDocument content, String kid) {
+        return content.getVerificationMethod().stream().filter(vm -> vm.getId().equals(kid))
+                .findFirst();
+    }
+
+    @NotNull
+    private Optional<VerificationMethod> firstVerificationMethod(DidDocument did) {
+        return did.getVerificationMethod().stream()
+                .filter(vm -> DidConstants.ALLOWED_VERIFICATION_TYPES.contains(vm.getType()))
+                .findFirst();
+    }
+}
```

Comment:
As discussed this implementation is almost similar to the `verifyJwtToken` of the `DecentralizedIdentityService`

Reply:
yes, there are many similarities, also regarding JWT validation and verification. I'd prefer to do the harmonization and clean up later, so that the PR doesn't get blown out of proportion.